### PR TITLE
Add sleep, passive mode for sensor

### PR DESCRIPTION
**Pull request by serunis**
Hello, I add sleep and passive mode for low power.
Tested with PMS7003 on windows10 Desktop, Raspberry pi 3B

Although I test only PMS7003, this would work for PMS5003 too.

Thank you

### CODE REVIEW

**Comment by FlorentinBulotAQ** [comment]
Dear Serunis,

Thanks a lot for your pull request. I think it makes much sense to add these features. 
I have seen on the other closed pull request that you found one other item to fix so let us know if that is still the case and when you have fixed and and we will review your changes.

With best wishes,
Florentin.

**Comment by serunis** [comment]
Hello Florentin,

I fixed the missing point, and updated with 'git push --force', so you cannot see the history,

I fixed the following flushing point:

    def read_in_passive(self, perform_flush=True):
        if perform_flush:
            self.serial.reset_input_buffer()  <-- 1. input flush here
        pms_cmd_read_in_passive = b'\x42\x4d\xe2\x00\x00\x01\x71'
        self.serial.write(pms_cmd_read_in_passive)
        self.serial.flush()
        ret = self.read(False)      <-- 2. call with False

My pull request is ready for review.
Thank you.

**Comment by pjb304** [comment]
Hi Serunis,
Many thanks for your contribution to this.
Flo is better placed to review this than I am so I've changed the reviewer over.

Phil

File: plantower/plantower.py [hunk 1]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+
+    def mode_change(self, mode=PMS_PASSIVE_MODE):
+        """
+            The default mode for the sensor is ACTIVE and whenever power OFF and ON (or sleep and wakeup)
+             the sensor starts in ACTIVE mode.
+            After mode changed, you need to wait some seconds for the sensor to get data again,
+            immediate reading sensor data gives you 0 values.
+        """
+
+        pms_cmd_changemode_passive = b'\x42\x4d\xe1\x00\x00\x01\x70'
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: I would suggest you put these two variables (pms_cmd_changemode_passive and pms_cmd_changemode_active) as global variable for consistency.

File: plantower/plantower.py [hunk 2]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+            self.serial.write(pms_cmd_changemode_active)
+
+        self.serial.flush()
+        time.sleep(0.2)
+        return
+
+    def read_in_passive(self, perform_flush=True):
+        if perform_flush:
+            self.serial.reset_input_buffer()  #Flush any data in the buffer
+        pms_cmd_read_in_passive = b'\x42\x4d\xe2\x00\x00\x01\x71'
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: Global variable

File: plantower/plantower.py [hunk 3]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+        time.sleep(0.5)  # wait sensor busy finished
+        return ret
+
+    def set_to_sleep(self, to_sleep=True):
+        """
+            This makes the sensor fan to stop.
+            From datasheet "Stable data should be got at least 30 seconds after the sensor wakeup
+            from the sleep mode because of the fan's performance."
+        """
+        pms_cmd_to_sleep = b'\x42\x4d\xe4\x00\x00\x01\x73'
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: Global variable

File: plantower/plantower.py [hunk 4]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+        pms_cmd_to_wakeup = b'\x42\x4d\xe4\x00\x01\x01\x74'
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: global variable

File: plantower/plantower.py [hunk 5]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+
+        pms_cmd_changemode_passive = b'\x42\x4d\xe1\x00\x00\x01\x70'
+        pms_cmd_changemode_active = b'\x42\x4d\xe1\x00\x01\x01\x71'
+
+        if mode == PMS_PASSIVE_MODE:
+            self.serial.write(pms_cmd_changemode_passive)
+        else:
+            self.serial.write(pms_cmd_changemode_active)
+
+        self.serial.flush()
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: I suggest using `self.serial.reset_input_buffer()` instead of `self.serial.flush()`

File: plantower/plantower.py [hunk 6]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+        time.sleep(0.2)
+        return
+
+    def read_in_passive(self, perform_flush=True):
+        if perform_flush:
+            self.serial.reset_input_buffer()  #Flush any data in the buffer
+        pms_cmd_read_in_passive = b'\x42\x4d\xe2\x00\x00\x01\x71'
+        self.serial.write(pms_cmd_read_in_passive)
+        self.serial.flush()
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: I suggest using `self.serial.reset_input_buffer()` instead of `self.serial.flush()`

File: plantower/plantower.py [hunk 7]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+        """
+        pms_cmd_to_sleep = b'\x42\x4d\xe4\x00\x00\x01\x73'
+        pms_cmd_to_wakeup = b'\x42\x4d\xe4\x00\x01\x01\x74'
+
+        if to_sleep:
+            self.serial.write(pms_cmd_to_sleep)
+        else:
+            self.serial.write(pms_cmd_to_wakeup)
+
+        self.serial.flush()
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: Use `self.serial.reset_input_buffer()` instead of `self.serial.flush()`

File: plantower/plantower.py [hunk 8]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+        time.sleep(2)  # number not specified in datasheet but sensor does not receive command for 2s.
+        return
+
+    def set_to_wakeup(self):
+        self.set_to_sleep(False)
+        return
+
+
+"""
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py:200]: This test script was very useful to test your code.

Could you put this test script into a separate file test.py ?

File: plantower/plantower.py [hunk 9]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+        else:
+            self.serial.write(pms_cmd_to_wakeup)
+
+        self.serial.flush()
+        time.sleep(2)  # number not specified in datasheet but sensor does not receive command for 2s.
+        return
+
+    def set_to_wakeup(self):
+        self.set_to_sleep(False)
+        return
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: I suggest to use `return True` or to delete the `return`

File: plantower/plantower.py [hunk 10]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+        pms_cmd_to_wakeup = b'\x42\x4d\xe4\x00\x01\x01\x74'
+
+        if to_sleep:
+            self.serial.write(pms_cmd_to_sleep)
+        else:
+            self.serial.write(pms_cmd_to_wakeup)
+
+        self.serial.flush()
+        time.sleep(2)  # number not specified in datasheet but sensor does not receive command for 2s.
+        return
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: I suggest to use `return True` or to delete the `return`

File: plantower/plantower.py [hunk 11]
```diff
@@ -18,6 +18,10 @@
 MSG_CHAR_1 = b'\x42' # First character to be recieved in a valid packet
 MSG_CHAR_2 = b'\x4d' # Second character to be recieved in a valid packet
 
+PMS_PASSIVE_MODE = 0
+PMS_ACTIVE_MODE = 1
+
+
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py:29]: I would suggest to add the commands as global variables for consistency

```PMS_CMD_CHANGEMODE_PASSIVE = b'\x42\x4d\xe1\x00\x00\x01\x70'
PMS_CMD_CHANGEMODE_ACTIVE = b'\x42\x4d\xe1\x00\x01\x01\x71'
PMS_CMD_READ_IN_PASSIVE = b'\x42\x4d\xe2\x00\x00\x01\x71'
PMS_CMD_TO_SLEEP = b'\x42\x4d\xe4\x00\x00\x01\x73'
PMS_CMD_TO_WAKEUP = b'\x42\x4d\xe4\x00\x01\x01\x74'
```

File: plantower/plantower.py (hunk 10 above)
Comment (FlorentinBulotAQ) [inline plantower/plantower.py]: I would suggest to add some logging, for example something like:

```
if to_sleep:
    self.logger.info("Sensor set to sleep")
else:
    self.logger.info("Sensor set to wakeup - 30 seconds required to obtain stable data")
```

File: plantower/plantower.py [hunk 12]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+    def read_in_passive(self, perform_flush=True):
+        if perform_flush:
+            self.serial.reset_input_buffer()  #Flush any data in the buffer
+        pms_cmd_read_in_passive = b'\x42\x4d\xe2\x00\x00\x01\x71'
+        self.serial.write(pms_cmd_read_in_passive)
+        self.serial.flush()
+        ret = self.read(False)
+        time.sleep(0.5)  # wait sensor busy finished
+        return ret
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py:179]: I suggest to add some logging, for example:

```
if mode == PMS_PASSIVE_MODE:
    self.logger.info("Sensor set in passive mode")
else:
   self.logger.info("Sensor set in active mode")
```
```

File: plantower/plantower.py [hunk 13]
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
...
+            self.serial.write(pms_cmd_changemode_passive)
+        else:
+            self.serial.write(pms_cmd_changemode_active)
+
+        self.serial.flush()
+        time.sleep(0.2)
+        return
+
+    def read_in_passive(self, perform_flush=True):
+        if perform_flush:
```
Comment (FlorentinBulotAQ) [inline plantower/plantower.py:172]: Could you add a short docstring at the beginning of this method?

**Review by FlorentinBulotAQ — COMMENTED** [review]
Dear @serunis,

I have tested your code and it works fine on a PMS5003. It was easy to read and to understand and all the methods you have implemented work.
I have just suggested a few minor correction for consistency with the rest of the code. But these are just suggestions so let me know what you think of them, I am happy to discuss it.


The only surprising behaviour is when the sensor is set to sleep, you can still read it using read_in_passive(). I think that is an issue with the sensor and not with the code, please see below:

```
In [15]: pms.set_to_sleep()

In [16]: pms.read()
---------------------------------------------------------------------------
PlantowerException                        Traceback (most recent call last)
<ipython-input-16-fc3fdc851aca> in <module>()
----> 1 pms.read()

/home/pi/plantower/plantower/plantower.py in read(self, perform_flush)
    138                     return PlantowerReading(recv) # convert to reading object
    139             #If the character isn't what we are expecting loop until timeout
--> 140         raise PlantowerException("No message recieved")
    141
    142     def mode_change(self, mode=PMS_PASSIVE_MODE):

PlantowerException: No message recieved
```

```
In [18]: print(pms.read_in_passive())
2020-01-21 11:59:25,5,5,6,6,6,6,948,248,31,2,0,0,

In [19]: print(pms.read_in_passive())
2020-01-21 11:59:29,5,5,6,6,6,6,948,248,31,2,0,0,

In [20]: print(pms.read_in_passive())
2020-01-21 11:59:32,5,5,6,6,6,6,948,248,31,2,0,0,

In [21]: print(pms.read_in_passive())
2020-01-21 11:59:36,5,5,6,6,6,6,948,248,31,2,0,0,

In [22]: print(pms.read_in_passive())
2020-01-21 11:59:39,5,5,6,6,6,6,948,248,31,2,0,0,

```

The buffer does not seem to be cleared and the sensor keeps returning the same data. Something to keep in mind when using the sensor.

Again thanks a lot for your pull request, all these changes are very to the point and very usefull to improve this library.

With best wishes,
Florentin.

File: plantower/plantower.py (hunk 13 above)
Comment (serunis) [reply plantower/plantower.py:172]: ok

File: plantower/plantower.py (hunk 11 above)
Comment (serunis) [reply plantower/plantower.py:29]: ok

File: plantower/plantower.py (hunk 12 above)
Comment (serunis) [reply plantower/plantower.py:179]: ok

File: plantower/plantower.py (hunk 8 above)
Comment (serunis) [reply plantower/plantower.py:200]: I don't know how to make test codes in the pythonic way.
I think, it is good for you to proceed.

**Comment by serunis** [comment]
Hello @FlorentinBulotUoS,
I updated code by your comment.
The usage of self.serial.flush() is intentional. I belive, it makes cpu to wait until tx finished.
and I think it is good for you to make test codes. I do not have a good idea of building test codes.

**Comment by serunis** [comment]
>The only surprising behaviour is when the sensor is set to sleep, you can still read it using read_in_passive().

Oh.. I did not know that, funny.

When I measure the power consumption of the sensor in sleep, it was way high for the controller in sleep.
I believe, the sleep mode of the sensor only meant the fan stops.

I originally made my own code for arduino, and I have made a portable outdoor wireless dust sensor solution. however, power comsumption in sleep mode was too higher than I expected, and I change my mind to make it indoor solution with Raspberry pi. 

Thanks to your library, I was happy that I do not have to parse the data structure in python again.
So I commit my impovement to your library.
Since an indoor solution still needs the sensor fan to stop due to its noise, a sleep mode is needed.

Actually this is my first github pull request, quite interesting!

Thank you for review.

File: plantower/plantower.py (hunk 8 above)
Comment (FlorentinBulotAQ) [reply plantower/plantower.py:200]: Ok, I will do it

**Review by FlorentinBulotAQ — APPROVED** [review]
Thanks a lot for making all these changes!

And I agree with you serial.flush() makes more sense in that case than serial.reset_input_buffer()

**Comment by FlorentinBulotAQ** [comment]
Hi @serunis 

I would like to credit you in the README.md file, so far I have added your name as serunis. Let me know if you would like to use a different name.

With best wishes,
Florentin.

**Comment by serunis** [comment]
serunis is good for me
Thank you

**Pull request merged**